### PR TITLE
 Add XML code generation

### PR DESCRIPTION
**Pull request by SupunS**

### CODE REVIEW

File: modules/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/XMLElementLiteralNode.java [hunk 1]
```diff
@@ -17,7 +17,11 @@
 */
 package org.ballerinalang.model.tree.expressions;
 
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.BSymbol;
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/XMLElementLiteralNode.java]: Don't use implementation classes in the API? unless there is a valid reason.

File: modules/ballerina-lang/src/main/java/org/ballerinalang/model/tree/statements/XMLNSDeclStatementNode.java [hunk 2]
```diff
@@ -18,10 +18,13 @@
 package org.ballerinalang.model.tree.statements;
 
 import org.ballerinalang.model.tree.XMLNSDeclarationNode;
+import org.wso2.ballerinalang.compiler.tree.BLangXMLNS;
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/ballerinalang/model/tree/statements/XMLNSDeclStatementNode.java]: This is not correct.

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 3]
```diff
@@ -489,31 +489,42 @@ public void visit(BLangXMLQName xmlQName) {
 
     @Override
     public void visit(BLangXMLAttribute xmlAttribute) {
+        xmlAttribute.name = rewrite(xmlAttribute.name);
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Use rewriteExpr() for all the expressions

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 4]
```diff
@@ -429,6 +483,14 @@ public void defineNode(BLangNode node, SymbolEnv env) {
         this.env = prevEnv;
     }
 
+    public void defineSymbol(DiagnosticPos pos, BSymbol symbol, SymbolEnv env) {
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Can we use the existing defineSymbol method?

Comment (sameera-jayasoma) [reply modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Btw, you need to check for duplicates before defining a symbol

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java [hunk 5]
```diff
@@ -251,6 +265,31 @@ public BSymbol lookupMemberSymbol(Scope scope, Name name, int expSymTag) {
...
+     * @param env Environment to get the visible namespaces
+     * @return Map of namespace symbols visible to the given environment
+     */
+    public Map<Name, BXMLNSSymbol> resolveAllNamespaces(SymbolEnv env) {
+        Map<Name, BXMLNSSymbol> namespaces = new HashMap<Name, BXMLNSSymbol>();
+        addNamespacesInScope(namespaces, env);
+        return namespaces;
+    }
+
+    private void addNamespacesInScope(Map<Name, BXMLNSSymbol> namespaces, SymbolEnv env) {
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java]: Move the private method to the private methods section.

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/symbols/SymTag.java [hunk 6]
```diff
@@ -39,5 +39,5 @@
     public static final int ANNOTATION = 1 << 13;
     public static final int ANNOTATION_ATTRIBUTE = 1 << 14;
     public static final int ERROR = 1 << 15;
-
+    public static final int XMLNS = 1 << 16;
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/symbols/SymTag.java]: ERROR should be the last one in the list.

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/TypeTags.java [hunk 7]
```diff
@@ -43,6 +43,7 @@
     public static final int NONE = INVOKABLE + 1;
     public static final int VOID = NONE + 1;
     public static final int ERROR = VOID + 1;
+    public static final int XMLNS = ERROR + 1;
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/TypeTags.java]: Error should be the last one

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 4 above)
Comment (SupunS) [reply modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Removed the method.

**Pull request merged**